Mathlib/RingTheory/NonZeroDivisors.lean

```diff
@@ -0,0 +1,189 @@
+/-
+Copyright (c) 2020 Kenny Lau. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenny Lau, Devon Tuma
+
+! This file was ported from Lean 3 source module ring_theory.non_zero_divisors
+! leanprover-community/mathlib commit 1126441d6bccf98c81214a0780c73d499f6721fe
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.GroupTheory.Submonoid.Operations
+import Mathlib.GroupTheory.Submonoid.Membership
+
+/-!
+# Non-zero divisors
+
+-- porting note: should this instead be called `NonZeroDivisors`? capitalisation looks weird...
+In this file we define the submonoid `nonZeroDivisors` of a `MonoidWithZero`.
+
+## Notations
+
+This file declares the notation `R⁰` for the submonoid of non-zero-divisors of `R`,
+in the locale `nonZeroDivisors`.
+
+Use the statement `open nonZeroDivisors` to access this notation in your own code.
+
+-/
+
+
+section nonZeroDivisors
+
+/-- The submonoid of non-zero-divisors of a `MonoidWithZero` `R`. -/
+def nonZeroDivisors (R : Type _) [MonoidWithZero R] : Submonoid R
+    where
+  carrier := { x | ∀ z, z * x = 0 → z = 0 }
+  one_mem' _ hz := by rwa [mul_one] at hz
+  mul_mem' hx₁ hx₂ _ hz := by
+    rw [← mul_assoc] at hz
+    exact hx₁ _ (hx₂ _ hz)
+#align non_zero_divisors nonZeroDivisors
+
+-- mathport name: non_zero_divisors
+/-- The notation for the submonoid of non-zerodivisors. -/
+scoped[nonZeroDivisors] notation:9000 R "⁰" => nonZeroDivisors R
+
+open nonZeroDivisors
+
+variable {M M' M₁ R R' F : Type _} [MonoidWithZero M] [MonoidWithZero M'] [CommMonoidWithZero M₁]
+  [Ring R] [CommRing R']
+
+theorem mem_nonZeroDivisors_iff {r : M} : r ∈ M⁰ ↔ ∀ x, x * r = 0 → x = 0 := Iff.rfl
+#align mem_non_zero_divisors_iff mem_nonZeroDivisors_iff
+
+theorem mul_right_mem_nonZeroDivisors_eq_zero_iff {x r : M} (hr : r ∈ M⁰) : x * r = 0 ↔ x = 0 :=
+  ⟨hr _, by simp (config := { contextual := true })⟩
+#align mul_right_mem_non_zero_divisors_eq_zero_iff mul_right_mem_nonZeroDivisors_eq_zero_iff
+-- porting note: what should coe theorems be called?
```

Comment:
We just use `coe` in the name. If we want to change it later, we'll do a refactor.
```suggestion

```